### ISSUE TITLE
[executor] parallelizing reads in the execution engine

### ISSUE THREAD

**Issue by wlawt** [ava-labs/hypersdk#701]
See below

Prereq: #693 
Related: #709

**Comment by wlawt** [comment]
Some thoughts:

Would it make sense to bundle RWrite and Write into the same set? In the current separation of StateKeys (Read/Write/RWrite), the `Write` set almost behaves like an `Alloc`, in the sense that all intended `Write`s end up becoming `RWrite`. This is because the key we want to write already exists, so we needed to get the value before writing (to my understanding, this is how it's currently implemented). In every `Action` example in TokenVM and MorpheusVM, almost all StateKeys are `RWrite`.

A follow-up thought to this would be handling dependencies. A naive approach would be that `RWrite` keys acquire an exclusive lock that allows them to read & write to that key, much like a `Write` key. However, given that we support the `RWrite` mode, it would be smarter to allocate shared locks on the reads and only acquire an exclusive lock on a write.

**Comment by wlawt** [comment]
Ways transactions can be parallelized:

- If transactions have same reads, different writes
- If transactions have different reads, different writes

**Comment by wlawt** [comment]
Parts of the execution engine that need to be reworked / thought about:

- Having the executor keep track of which keys are read and write
- Having an updated criteria of when to add to dependency and blocking
- How to tell if a transaction can be ran in parallel

**Comment by wlawt** [comment]
Any good examples to test for performance improvement?

**Comment by wlawt** [comment]
Some examples of parallelizing reads:

Example 1: Two txns minting an NFT, if the price of X (some arbitrary token) is greater than a certain price
Transaction 1:

Price of X (a read, assume the key already exists)
Create a new NFT (a write)
Sending from Actor 1 (a RW)
Transaction 2:

Price of X (a read, assume the key already exists)
Create a new NFT (a write)
Sending from Actor 2 (a RW)
Here, we can parallelize T1 and T2 since there are no conflicts.

**Comment by patrick-ogrady** [comment]
> Some examples of parallelizing reads:
> 
> Example 1: Two txns minting an NFT, if the price of X (some arbitrary token) is greater than a certain price Transaction 1:
> 
> Price of X (a read, assume the key already exists) Create a new NFT (a write) Sending from Actor 1 (a RW) Transaction 2:
> 
> Price of X (a read, assume the key already exists) Create a new NFT (a write) Sending from Actor 2 (a RW) Here, we can parallelize T1 and T2 since there are no conflicts.

One thing I've always wanted to figure out is how to track "max supply" concurrently (as that will likely be a topic).

Or, it would be interesting to pre-create all "IDs" and then allow people to just claim them (so no contention on the supply key).

**Comment by richardpringle** [comment]
I just wrote out a whole thing and somehow deleted it.

TL;DR (because the original content is actually gone) 

NFT-max-supply is just for caps, right?

You could have *limited*-determinism in calculating the supply. Consolidate the shared count every X blocks... so in the C(x - 1) blocks, you could potentially pump the supply, but it would be cut off after that and is still limited by the throughput of the chain. 

I was also thinking you could just have queues and have users determine which queue they want to try and claim their ID from... but then I thought, someone would just create a program to do this for them and bottleneck the whole NFT (even if it cost them more).

Won't that always be a problem? Or is there a fee mechanism that will prevent this? 

As for fungible tokens, I'm guessing that in a typical use-case that cares about throughput, they would only care about transfer throughput, not mind/burn throughput.

**Comment by wlawt** [comment]
Created an issue #712 to address the comments around Counters, as it will also be a separate follow-up PR. This current issue will just focus on the basic issue of updating the Execution Engine to be able to handle read-keys and write/allocate-keys
